packages/dev/core/src/FlowGraph/Blocks/Execution/ControlFlow/flowGraphMultiGateBlock.ts

```diff
@@ -0,0 +1,89 @@
+import type { FlowGraphContext } from "../../../flowGraphContext";
+import type { FlowGraphDataConnection } from "../../../flowGraphDataConnection";
+import { FlowGraphExecutionBlock } from "../../../flowGraphExecutionBlock";
+import { RichTypeNumber } from "../../../flowGraphRichTypes";
+import type { FlowGraphSignalConnection } from "../../../flowGraphSignalConnection";
+
+export interface IFlowGraphMultiGateBlockConfiguration {
+    numberOutputFlows: number;
+    isRandom?: boolean;
+    loop?: boolean;
+    startIndex?: number;
+}
+/**
+ * @experimental
+ * A block that has an input flow and routes it to any potential output flows, randomly or sequentially
+ * @see https://docs.google.com/document/d/1MT7gL-IEn_PUw-4XGBazMxsyqsxqgAVGYcNeC4Cj_9Q/edit#heading=h.i2sn85fbjo60
+ */
+export class FlowGraphMultiGateBlock extends FlowGraphExecutionBlock {
+    private _config: IFlowGraphMultiGateBlockConfiguration;
+    public readonly reset: FlowGraphSignalConnection;
+    public readonly outFlows: FlowGraphSignalConnection[] = [];
+    public readonly currentIndex: FlowGraphDataConnection<number>;
+
+    constructor(config: IFlowGraphMultiGateBlockConfiguration) {
+        super();
+        this._config = {
+            isRandom: false,
+            loop: false,
+            startIndex: 0,
+            ...config,
+        };
+        this._config.startIndex = Math.max(0, Math.min(this._config.startIndex!, this._config.numberOutputFlows - 1));
+
+        for (let i = 0; i < this._config.numberOutputFlows; i++) {
+            this.outFlows.push(this._registerSignalOutput(`out${i}`));
+        }
+
+        this.reset = this._registerSignalInput("reset");
+        this.currentIndex = this._registerDataOutput("currentIndex", RichTypeNumber);
+    }
+
+    private _getUnusedIndexes(context: FlowGraphContext): number[] {
+        let result = [];

```

Comment:
this array can be reused instead of creating a new one, as it is only used inside the execute (and is recreated there as well, with a `filter` call